commons/all/src/main/java/org/infinispan/commons/util/SslContextFactory.java

```diff
@@ -204,4 +229,38 @@ private static void loadKeyStore(KeyStore ks, String keyStoreFileName, char[] ke
          Util.close(is);
       }
    }
+
+   public static Provider findProvider(String providerName, String serviceType, String algorithm) {
+      Provider[] providers = discoverSecurityProviders(Thread.currentThread().getContextClassLoader());
+      for (Provider provider : providers) {
+         if (providerName == null || providerName.equals(provider.getName())) {
+            Provider.Service providerService = provider.getService(serviceType, algorithm);
+            if (providerService != null) {
+               return provider;
+            }
+         }
+      }
+      return null;
+   }
+
+   public static Provider[] discoverSecurityProviders(ClassLoader classLoader) {
+      return PER_CLASSLOADER_PROVIDERS.computeIfAbsent(classLoader, cl -> {
+         Map<Class<? extends Provider>, Provider> providers = new LinkedHashMap<>();
+         for (Provider provider : Security.getProviders()) {
+            providers.put(provider.getClass(), provider);
+         }
+         ServiceLoader<Provider> loader = ServiceLoader.load(Provider.class, classLoader);
+         Iterator<Provider> iterator = loader.iterator();
+         for (; ; )
```

Comment:
just curious why not using single line 
`for (Iterator<Provider> iterator = ServiceLoader.load(Provider.class, classLoader).iterator(); iterator.hasNext(); )`

Reply:
Even better:
```java
for (Provider provider : ServiceLoader.load(Provider.class, classLoader))
```